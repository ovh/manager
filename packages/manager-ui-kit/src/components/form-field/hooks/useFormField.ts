import { useFormField } from '@ovhcloud/ods-react';

export { useFormField };
