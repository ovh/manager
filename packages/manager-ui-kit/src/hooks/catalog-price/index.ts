export { useCatalogPrice } from './useCatalogPrice';
export type { CatalogPriceOptions } from './useCatalog.props';
export { priceToUcent, priceFromUcent, convertHourlyPriceToMonthly } from './useCatalog.utils';
