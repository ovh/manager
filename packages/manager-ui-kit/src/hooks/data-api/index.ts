export * from './ports/use-data-api';
