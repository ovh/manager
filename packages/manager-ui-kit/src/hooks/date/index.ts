export * from './useDateFnsLocale';
export * from './useFormatDate';
