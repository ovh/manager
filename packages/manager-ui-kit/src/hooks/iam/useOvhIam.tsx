import { useMemo } from 'react';

import { keepPreviousData, useQuery } from '@tanstack/react-query';

import { apiClient } from '@ovh-ux/manager-core-api';

import { formatIamTagsFromResources } from '@/commons/utils/TagsHelper';
import { useDataApi } from '@/hooks/data-api/ports/useDataApi';
import { IAMType, IamCheckResponse, IamObject } from '@/hooks/iam/IAM.type';

export const fetchAuthorizationsCheck = async ({
  actions,
  urns,
}: {
  actions: string[];
  urns: string[];
}) => {
  const response = await apiClient.v2?.post(`/iam/authorization/check`, {
    actions,
    resourceURNs: urns,
  });
  return response?.data;
};

export function useAuthorizationsIam({ actions, urns }: IAMType) {
  const { data } = useQuery({
    queryKey: ['iam-authorization', urns, actions],
    queryFn: () => fetchAuthorizationsCheck({ actions: actions || [], urns: urns || [] }),
    enabled: (urns?.length || 0) > 0 && (actions?.length || 0) > 0,
    placeholderData: keepPreviousData,
  });

  return data || [];
}

export function getAuthorizationCheckUrl(urn: string) {
  return `/iam/resource/${encodeURIComponent(urn)}/authorization/check`;
}

export const fetchAuthorizationCheck = async (
  actions: string[],
  urn: string,
): Promise<IamCheckResponse> => {
  const response = await apiClient.v2?.post(getAuthorizationCheckUrl(urn), {
    actions,
  });
  return response?.data;
};

export function useAuthorizationIam(
  actions: string[],
  urn: string,
  isTrigger = true,
  // `isTrigger` is by default at true, to request at the init of the hook.
  // In the case of action menu, we want to trigger it, when user click on button
) {
  const { data, ...query } = useQuery({
    queryKey: [urn, actions],
    queryFn: () => fetchAuthorizationCheck(actions, urn),
    enabled: Boolean(urn && urn.length > 0 && actions && actions.length > 0 && isTrigger),
    placeholderData: keepPreviousData,
  });

  return {
    isAuthorized:
      !!data?.authorizedActions &&
      actions.every((action) => data?.authorizedActions?.includes(action)),
    data,
    ...query,
  };
}

export function useGetResourceTags({
  resourceType,
  enabled = true,
}: {
  resourceType?: string;
  enabled?: boolean;
}) {
  let route = '/iam/resource';

  if (resourceType) {
    route = `${route}?resourceType=${resourceType}`;
  }

  const {
    flattenData: resources,
    isError: isTagsError,
    isLoading: isTagsLoading,
  } = useDataApi<IamObject>({
    version: 'v2',
    iceberg: true,
    route,
    cacheKey: ['iam/resource', resourceType || ''],
    enabled,
    fetchAll: true,
  });

  const tags = useMemo(() => {
    if (!resources) {
      return [];
    }
    return formatIamTagsFromResources(resources);
  }, [resources]);

  return {
    tags,
    isError: isTagsError,
    isLoading: isTagsLoading,
  };
}
