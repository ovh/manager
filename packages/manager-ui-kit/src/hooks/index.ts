export { useBytes } from './bytes/useBytes';
export { useCatalogPrice } from './catalog-price/useCatalogPrice';
export { useDataApi } from './data-api/ports/useDataApi';
export { useColumnFilters } from './data-api/useColumnFilters';
export { useDatagridSearchParams } from './data-api/useDatagridSearchParams';
export { useDateFnsLocale } from './date/date-fns-locale/useDateFnsLocale';
export { DEFAULT_UNKNOWN_DATE_LABEL } from './date/date-formatter/FormatDate.type';
export { useFormatDate } from './date/date-formatter/useFormatDate';
export { useMe } from './me/useMe';
export { isLocalZone, getMacroRegion } from './region/Regions.utils';
export { useTranslatedMicroRegions } from './region/useTranslatedMicroRegions';

export * from './iam/IAM.type';
export * from './iam/useOvhIam';

export type { CatalogPriceOptions } from './catalog-price/Catalog.type';
export type { UseDataApiOptions, UseDataApiResult } from './data-api/ports/useDataApi.types';
export type { LOCALE_MAP } from './date/date-fns-locale/DateFnsLocale.constants';
export type { LocaleKey } from './date/date-fns-locale/DateFnsLocale.type';
export type { FormatDateOptions } from './date/date-formatter/FormatDate.type';
export type { IMe } from './me/IMe.type';

export {
  priceToUcent,
  priceFromUcent,
  convertHourlyPriceToMonthly,
} from './catalog-price/Catalog.utils';
