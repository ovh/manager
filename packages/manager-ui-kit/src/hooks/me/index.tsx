export { useMe } from './useMe';
export type { IMe } from './useMe.props';
