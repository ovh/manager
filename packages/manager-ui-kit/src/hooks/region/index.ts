export { useTranslatedMicroRegions } from './useTranslatedMicroRegions';
export { isLocalZone, getMacroRegion } from './useTranslatedMicroRegions.utils';
