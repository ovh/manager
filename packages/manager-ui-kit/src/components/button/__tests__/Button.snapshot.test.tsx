import { vitest } from 'vitest';
import type { MockInstance } from 'vitest';
import { TOOLTIP_POSITION } from '@ovhcloud/ods-react';
import { render } from '@/setupTest';
import { Button } from '../index';
import { useAuthorizationIam } from '../../../hooks/iam';

vitest.mock('../../../hooks/iam', () => ({
  useAuthorizationIam: vitest.fn().mockReturnValue({
    isAuthorized: true,
    isLoading: false,
    isFetched: true,
  }),
}));

const mockedHook = useAuthorizationIam as unknown as MockInstance;

describe('Button Snapshot Tests', () => {
  afterEach(() => {
    vitest.resetAllMocks();
  });

  describe('Authorized Button States', () => {
    it('should render basic button when authorized', () => {
      mockedHook.mockReturnValue({
        isAuthorized: true,
        isLoading: false,
        isFetched: true,
      });

      const { container } = render(<Button>Click me</Button>);

      expect(container).toMatchSnapshot();
    });

    it('should render button with IAM props when authorized', () => {
      mockedHook.mockReturnValue({
        isAuthorized: true,
        isLoading: false,
        isFetched: true,
      });

      const { container } = render(
        <Button
          iamActions={['test:action']}
          urn="urn:test:resource"
          displayTooltip={true}
          isIamTrigger={true}
        >
          Authorized Button
        </Button>,
      );

      expect(container).toMatchSnapshot();
    });

    it('should render button with ODS button props when authorized', () => {
      mockedHook.mockReturnValue({
        isAuthorized: true,
        isLoading: false,
        isFetched: true,
      });

      const { container } = render(
        <Button
          iamActions={['test:action']}
          urn="urn:test:resource"
          displayTooltip={true}
          isIamTrigger={true}
        >
          Full Featured Button
        </Button>,
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('Unauthorized Button States', () => {
    it('should render disabled button with tooltip when unauthorized', () => {
      mockedHook.mockReturnValue({
        isAuthorized: false,
        isLoading: false,
        isFetched: true,
      });

      const { container } = render(
        <Button
          iamActions={['test:action']}
          urn="urn:test:resource"
          displayTooltip={true}
          tooltipPosition={TOOLTIP_POSITION.bottom}
        >
          Unauthorized Button
        </Button>,
      );

      expect(container).toMatchSnapshot();
    });

    it('should render disabled button without tooltip when unauthorized', () => {
      mockedHook.mockReturnValue({
        isAuthorized: false,
        isLoading: false,
        isFetched: true,
      });

      const { container } = render(
        <Button
          iamActions={['test:action']}
          urn="urn:test:resource"
          displayTooltip={false}
        >
          Unauthorized Button No Tooltip
        </Button>,
      );

      expect(container).toMatchSnapshot();
    });

    it('should render disabled button with custom tooltip position', () => {
      mockedHook.mockReturnValue({
        isAuthorized: false,
        isLoading: false,
        isFetched: true,
      });

      const { container } = render(
        <Button
          iamActions={['test:action']}
          urn="urn:test:resource"
          displayTooltip={true}
          tooltipPosition={TOOLTIP_POSITION.top}
        >
          Unauthorized Button Top Tooltip
        </Button>,
      );

      expect(container).toMatchSnapshot();
    });

    it('should render disabled button with all ODS props when unauthorized', () => {
      mockedHook.mockReturnValue({
        isAuthorized: false,
        isLoading: false,
        isFetched: true,
      });

      const { container } = render(
        <Button
          iamActions={['test:action']}
          urn="urn:test:resource"
          displayTooltip={true}
          tooltipPosition={TOOLTIP_POSITION.right}
        >
          Unauthorized Full Featured Button
        </Button>,
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('Edge Cases', () => {
    it('should render button when no IAM props provided', () => {
      mockedHook.mockReturnValue({
        isAuthorized: true,
        isLoading: false,
        isFetched: true,
      });

      const { container } = render(<Button>No IAM Button</Button>);

      expect(container).toMatchSnapshot();
    });

    it('should render button when only iamActions provided', () => {
      mockedHook.mockReturnValue({
        isAuthorized: true,
        isLoading: false,
        isFetched: true,
      });

      const { container } = render(
        <Button iamActions={['test:action']}>IAM Actions Only Button</Button>,
      );

      expect(container).toMatchSnapshot();
    });

    it('should render button when only urn provided', () => {
      mockedHook.mockReturnValue({
        isAuthorized: true,
        isLoading: false,
        isFetched: true,
      });

      const { container } = render(
        <Button urn="urn:test:resource">URN Only Button</Button>,
      );

      expect(container).toMatchSnapshot();
    });

    it('should render button with JSX children', () => {
      mockedHook.mockReturnValue({
        isAuthorized: true,
        isLoading: false,
        isFetched: true,
      });

      const { container } = render(
        <Button iamActions={['test:action']} urn="urn:test:resource">
          <span>JSX Children Button</span>
        </Button>,
      );

      expect(container).toMatchSnapshot();
    });

    it('should render button with isIamTrigger set to false', () => {
      mockedHook.mockReturnValue({
        isAuthorized: true,
        isLoading: false,
        isFetched: true,
      });

      const { container } = render(
        <Button
          iamActions={['test:action']}
          urn="urn:test:resource"
          isIamTrigger={false}
        >
          Non-IAM Trigger Button
        </Button>,
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('Loading States', () => {
    it('should render button when IAM is loading', () => {
      mockedHook.mockReturnValue({
        isAuthorized: false,
        isLoading: true,
        isFetched: false,
      });

      const { container } = render(
        <Button iamActions={['test:action']} urn="urn:test:resource">
          Loading IAM Button
        </Button>,
      );

      expect(container).toMatchSnapshot();
    });

    it('should render button when IAM is loading and authorized', () => {
      mockedHook.mockReturnValue({
        isAuthorized: true,
        isLoading: true,
        isFetched: false,
      });

      const { container } = render(
        <Button iamActions={['test:action']} urn="urn:test:resource">
          Loading Authorized Button
        </Button>,
      );

      expect(container).toMatchSnapshot();
    });
  });
});
