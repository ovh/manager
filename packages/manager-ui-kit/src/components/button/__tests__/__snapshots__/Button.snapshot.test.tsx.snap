// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Snapshot Tests > Authorized Button States > should render basic button when authorized 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49"
    data-ods="button"
    data-testid="manager-button"
    type="button"
  >
    Click me
  </button>
</div>
`;

exports[`Button Snapshot Tests > Authorized Button States > should render button with IAM props when authorized 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49"
    data-ods="button"
    data-testid="manager-button"
    type="button"
  >
    Authorized Button
  </button>
</div>
`;

exports[`Button Snapshot Tests > Authorized Button States > should render button with ODS button props when authorized 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49"
    data-ods="button"
    data-testid="manager-button"
    type="button"
  >
    Full Featured Button
  </button>
</div>
`;

exports[`Button Snapshot Tests > Edge Cases > should render button when no IAM props provided 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49"
    data-ods="button"
    data-testid="manager-button"
    type="button"
  >
    No IAM Button
  </button>
</div>
`;

exports[`Button Snapshot Tests > Edge Cases > should render button when only iamActions provided 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49"
    data-ods="button"
    data-testid="manager-button"
    type="button"
  >
    IAM Actions Only Button
  </button>
</div>
`;

exports[`Button Snapshot Tests > Edge Cases > should render button when only urn provided 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49"
    data-ods="button"
    data-testid="manager-button"
    type="button"
  >
    URN Only Button
  </button>
</div>
`;

exports[`Button Snapshot Tests > Edge Cases > should render button with JSX children 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49"
    data-ods="button"
    data-testid="manager-button"
    type="button"
  >
    <span>
      JSX Children Button
    </span>
  </button>
</div>
`;

exports[`Button Snapshot Tests > Edge Cases > should render button with isIamTrigger set to false 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49"
    data-ods="button"
    data-testid="manager-button"
    type="button"
  >
    Non-IAM Trigger Button
  </button>
</div>
`;

exports[`Button Snapshot Tests > Loading States > should render button when IAM is loading 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49 _tooltip-trigger_wxukn_2"
    data-ods="tooltip-trigger"
    data-part="trigger"
    data-scope="tooltip"
    data-state="closed"
    data-testid="manager-button-tooltip"
    dir="ltr"
    disabled=""
    id="tooltip::r3::trigger"
    role="button"
    tabindex="0"
    type="button"
  >
    Loading IAM Button
  </button>
</div>
`;

exports[`Button Snapshot Tests > Loading States > should render button when IAM is loading and authorized 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49"
    data-ods="button"
    data-testid="manager-button"
    type="button"
  >
    Loading Authorized Button
  </button>
</div>
`;

exports[`Button Snapshot Tests > Unauthorized Button States > should render disabled button with all ODS props when unauthorized 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49 _tooltip-trigger_wxukn_2"
    data-ods="tooltip-trigger"
    data-part="trigger"
    data-scope="tooltip"
    data-state="closed"
    data-testid="manager-button-tooltip"
    dir="ltr"
    disabled=""
    id="tooltip::r2::trigger"
    role="button"
    tabindex="0"
    type="button"
  >
    Unauthorized Full Featured Button
  </button>
</div>
`;

exports[`Button Snapshot Tests > Unauthorized Button States > should render disabled button with custom tooltip position 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49 _tooltip-trigger_wxukn_2"
    data-ods="tooltip-trigger"
    data-part="trigger"
    data-scope="tooltip"
    data-state="closed"
    data-testid="manager-button-tooltip"
    dir="ltr"
    disabled=""
    id="tooltip::r1::trigger"
    role="button"
    tabindex="0"
    type="button"
  >
    Unauthorized Button Top Tooltip
  </button>
</div>
`;

exports[`Button Snapshot Tests > Unauthorized Button States > should render disabled button with tooltip when unauthorized 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49 _tooltip-trigger_wxukn_2"
    data-ods="tooltip-trigger"
    data-part="trigger"
    data-scope="tooltip"
    data-state="closed"
    data-testid="manager-button-tooltip"
    dir="ltr"
    disabled=""
    id="tooltip::r0::trigger"
    role="button"
    tabindex="0"
    type="button"
  >
    Unauthorized Button
  </button>
</div>
`;

exports[`Button Snapshot Tests > Unauthorized Button States > should render disabled button without tooltip when unauthorized 1`] = `
<div>
  <button
    class="_button_1koyw_2 _button--primary_1koyw_100 _button--md_1koyw_42 _button--default_1koyw_49"
    data-ods="button"
    data-testid="manager-button-without-tooltip"
    disabled=""
    type="button"
  >
    Unauthorized Button No Tooltip
  </button>
</div>
`;
